src/hotspot/share/opto/vectorIntrinsics.cpp

```diff
@@ -1741,32 +1741,41 @@ bool LibraryCallKit::inline_vector_compare() {
 }
 
 // public static
-// <V extends Vector, Sh extends Shuffle>
-//  V rearrangeOp(Class<V> vectorClass, Class<Sh> shuffleClass, Class< ? > elementType, int vlen,
-//    V v1, Sh sh,
-//    VectorSwizzleOp<V, Sh, S, E> defaultImpl) { ...
-
+// <V extends Vector<E>,
+//     Sh extends VectorShuffle<E>,
+//     M extends VectorMask<E>,
+//     E>
+// V rearrangeOp(Class<? extends V> vectorClass, Class<Sh> shuffleClass, Class<M> maskClass, Class<?> elementType, int vlen,
+//               V v1, Sh sh, M m,
+//               VectorRearrangeOp<V,Sh,M,E> defaultImpl)
 bool LibraryCallKit::inline_vector_rearrange() {
   const TypeInstPtr* vector_klass  = gvn().type(argument(0))->isa_instptr();
   const TypeInstPtr* shuffle_klass = gvn().type(argument(1))->isa_instptr();
-  const TypeInstPtr* elem_klass    = gvn().type(argument(2))->isa_instptr();
-  const TypeInt*     vlen          = gvn().type(argument(3))->isa_int();
+  const TypeInstPtr* mask_klass    = gvn().type(argument(2))->isa_instptr();
+  const TypeInstPtr* elem_klass    = gvn().type(argument(3))->isa_instptr();
+  const TypeInt*     vlen          = gvn().type(argument(4))->isa_int();
 
-  if (vector_klass == NULL || shuffle_klass == NULL || elem_klass == NULL || vlen == NULL) {
+  if (vector_klass == NULL || shuffle_klass == NULL || elem_klass == NULL || mask_klass == NULL || vlen == NULL) {
```

Comment:
It allows the `mask_klass` to be `NULL` for non-masked rearrange, right? If so I think it's better to only check the `mask_klass` for `is_masked_op`.